observer/probers/tls/tls.go

```diff
@@ -0,0 +1,193 @@
+func (p TLSProbe) getInsecureCertInfo() (string, string, time.Duration, error) {
+	conf := &tls.Config{InsecureSkipVerify: true}
+	conn, err := tls.Dial("tcp", p.url+":443", conf)
+	if err != nil {
+		return "", "", 0, err
+	}
+	peers := conn.ConnectionState().PeerCertificates
+	expiryDate := peers[0].NotAfter
+
+	for i := 0; i < len(peers)-1; i++ {
+		certIssuer, issuerSubject := peers[i].Issuer, peers[i+1].Subject
+		if (certIssuer.CommonName != issuerSubject.CommonName) || (certIssuer.Organization[0] != issuerSubject.Organization[0]) {
+			err = fmt.Errorf(reasonToString[issuerVerifyFailed])
+		}
+	}
+
+	root := peers[len(peers)-1].Issuer
+	return root.Organization[0], root.CommonName, time.Until(expiryDate), err
+}
+
+// Check if the root of the leaf's certificate is the one we expect.
+func (p TLSProbe) isRootExpected(rootO, rootCN string) bool {
+	if p.rootOrg == rootO && p.rootCN == rootCN {
+		return true
+	}
+	return false
+}
+
+// Export time to (or since) expiration (in seconds) and reason (with
+// corresponding badOutcome label) to Prometheus.
+func (p TLSProbe) exportMetrics(expiryTime time.Duration, reason reason) {
+	p.certExpiry.WithLabelValues(p.url).Set(expiryTime.Seconds())
+	p.certExpiry.WithLabelValues(p.url, badOutcomeError{reason}.Error()).Set(float64(reason))
+}
+
+// Probe performs the configured TLS protocol. Return true if both root AND
+// response are the expected values, otherwise false. Export time to cert expiry
+// and reason as Prometheus metrics.
+func (p TLSProbe) Probe(timeout time.Duration) (bool, time.Duration) {
+	start := time.Now()
+	conn, secure_err := tls.Dial("tcp", p.url+":443", &tls.Config{})
+	if secure_err != nil {
+		// TLS connection failed, so try insecure connection instead.
+		rootO, rootCN, timeUntilExpiry, err := p.getInsecureCertInfo()
+		// Insecure connection successful, but certificate chain is invalid.
+		if errors.Is(err, fmt.Errorf(reasonToString[issuerVerifyFailed])) {
+			p.exportMetrics(timeUntilExpiry, issuerVerifyFailed)
+			return false, time.Since(start)
+		}
+		// Insecure connection failed.
+		if err != nil {
+			p.exportMetrics(timeUntilExpiry, internalError)
+			return false, time.Since(start)
+		}
+		// Insecure connection successful and chain is valid, so check if
+		// certificate is and should be expired.
+		if p.response == "expired" {
+			certInvalidErr := x509.CertificateInvalidError{}
+			p.exportMetrics(timeUntilExpiry, none)
+			return (errors.As(secure_err, &certInvalidErr) && certInvalidErr.Reason == x509.Expired) && p.isRootExpected(rootO, rootCN), time.Since(start)
+		}
+		// None of the above cases. Secure TLS connection failed and certificate
+		// should not be expired, so return false.
+		p.exportMetrics(timeUntilExpiry, internalError)
+		return false, time.Since(start)
+	}
+	// Certificate has been validated. Now check if it's revoked.
+	defer conn.Close()
+	chains := conn.ConnectionState().VerifiedChains
```

Comment:
I wouldn't use VerifiedChains at all. The certs present here will change depending on operating system, trust store, which way the wind blows, time of day, etc :D Instead, let's just look at the certs actually presented by the site: PeerCertificates.